Scripts/models/linear.py

```diff
@@ -0,0 +1,105 @@
+import numpy
+import pandas
+import math
+import parameters
+
+
+class LinearModel(object):
+    def __init__(self, zone_data_base, zone_data_forecast, bounds, resultdata):
+        """Initialize a linear model.
+
+        Parameters
+        ----------
+        zone_data : ZoneData
+            A ZoneData object defining the input data of zones.
+        bounds : slice
+            Defines the area on which the model is predicting to (usually the
+            metropolitan area).
+        """        
+        self.zone_data = zone_data_forecast
+        self.zone_data_base = zone_data_base
+        self.bounds = bounds
+        self.resultdata = resultdata
+
+    def _add_constant(self, prediction, b):
+        try: # If only one parameter
+            prediction += b
+        except ValueError: # Separate params for cap region and surrounding
+            k = self.zone_data.first_surrounding_zone
+            prediction[:k] += b[0]
+            prediction[k:] += b[1]
+        return prediction
+
+    def _add_zone_terms(self, prediction, b, generation=False):
+        zdata = self.zone_data
+        for i in b:
+            try: # If only one parameter
+                prediction += b[i] * zdata.get_data(i, self.bounds, generation)
+            except ValueError: # Separate params for cap region and surrounding
+                k = self.zone_data.first_surrounding_zone
+                data_capital_region = zdata.get_data(
+                    i, self.bounds, generation, zdata.CAPITAL_REGION)
+                data_surrounding = zdata.get_data(
+                    i, self.bounds, generation, zdata.SURROUNDING_AREA)
+                prediction[:k] += b[i][0] * data_capital_region
+                prediction[k:] += b[i][1] * data_surrounding
+        return prediction
+
+    def _add_log_zone_terms(self, prediction, b, generation=False):
+        zdata = self.zone_data
+        for i in b:
+            prediction += b[i] * numpy.log(zdata.get_data(i, self.bounds,
+                generation))
```

Comment:
Share of detached houses from self.zonedata include zeros (full model), so this would calculate log(0), which causes a warning. Could there be clip in zonedata.py?

Reply:
But detached houses variable is not log (?)

Reply:
But detached houses variable is not log (?)